hbase_sql/src/main/java/com/splicemachine/stream/RemoteQueryClientImpl.java

```diff
@@ -276,4 +278,10 @@ public void close() throws Exception {
         if (olapFuture != null)
             olapFuture.cancel(false);
     }
+
+    @SuppressFBWarnings(value = "RV_RETURN_VALUE_IGNORED", justification = "intended")
+    public Exception getException() throws InterruptedException {
+        olapFutureCallbackInvoked.await(5, TimeUnit.SECONDS);
```

Comment:
another way, where failure is updated, is exceptionCaught, which handles channel exceptions. Not sure if it can happen, but failure can be set already from the channel before olapFutureCallbackInvoked downs. To avoid additional waiting, maybe it's possible to check streamListener.getFailure() != null before olapFutureCallbackInvoked.await and return immediately or wait if still null.

Reply:
And that's exactly the point of this fix. If some other thread set the `failure` field before `olapFutureCallbackInvoked` count down to 0, that means the exception is not the real exception from olap side and we should wait for olapFuture callback to be invoked.

Point is, as long as `olapFutureCallbackInvoked` is not 0, we wait and don't trust the exception set to `failure`. We use that exception only when 1) olap callback is executed but no exception, or 2) olap callback timed out in 5 seconds.

Reply:
thank you, got it